### ISSUE TITLE
Missing route rules for pods

### ISSUE THREAD

**Issue by mateuszkwiatkowski** [projectcalico/calico#6169]
I've recently upgraded our Kubernetes cluster (rke) from 1.20.x to 1.22.9 and calico from 3.13.3 to 3.21.1 After some time new pods can't get working networking. With tcpdump I discovered that packets from node to pods are being routed via default gateway and liveness probes started to fail.

Check route for newly created pod:
```
# ip r get 10.255.48.199
10.255.48.199 via 10.35.28.1 dev bond0 src 10.35.28.29 uid 0
    cache
```

Check route for old pod:
```
# ip r get 10.255.48.205
10.255.48.205 dev calid17eea0b2d9 src 10.35.28.29 uid 0
    cache
```

I came up with quick one-liner to find all broken pods on node:
```
for IP in $(KUBECONFIG=/etc/kubernetes/ssl/kubecfg-kube-node.yaml kubectl get pods -A -o jsonpath='{.items[?(@.spec.nodeName=="NODENAME")].status.podIP}'); do ip r get $IP ; done | grep bond0
```

Next, I found out that neither route nor iptables rules are being created for new pods:
```
# ip link sh calie341b9aad75
13544: calie341b9aad75@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP mode DEFAULT group default
    link/ether ee:ee:ee:ee:ee:ee brd ff:ff:ff:ff:ff:ff link-netnsid 43
# ip r | grep calie341b9aad75
[empty output]
# iptables -L -nvx | grep calie341b9aad75
[empty output]
```

What's more, deleting pod results in stale iptables rules being left:
```
# ip link | grep cali40641bddbbb
[empty output]
# iptables -L -nvx | grep cali40641bddbbb
   97998 13956391 cali-fw-cali40641bddbbb  all  --  cali40641bddbbb *       0.0.0.0/0            0.0.0.0/0           [goto]  /* cali:lsknnSn91np3vngX */
Chain cali-fw-cali40641bddbbb (1 references)
       0        0 cali-tw-cali40641bddbbb  all  --  *      cali40641bddbbb  0.0.0.0/0            0.0.0.0/0           [goto]  /* cali:BdgMaIg5bESdEkks */
Chain cali-tw-cali40641bddbbb (1 references)
```

Restarting node or calico process fixes the problem. 
Issue is somewhat similar to https://github.com/projectcalico/calico/issues/5760 but our cluster is small (only 5 worker nodes, 3 controlplane nodes).  I also checked that we only have rules in itpables-legacy backend - iptables-nft is empty. Iptables backend is managed with `update-alternatives(1)` and is left in default setting:
```
# ls -l $(which iptables)
lrwxrwxrwx 1 root root 26 Jan 13  2021 /usr/sbin/iptables -> /etc/alternatives/iptables
# ls -l /etc/alternatives/iptables
lrwxrwxrwx 1 root root 25 Jan 13  2021 /etc/alternatives/iptables -> /usr/sbin/iptables-legacy
```

## Expected Behavior
New pods should get working networking all the time. Iptables rules should be deleted after pod is removed.

## Current Behavior
Afrter some time new pods have broken networking. Stale iptables rules are left after deleted pods.

## Possible Solution
Workaround: restart calico or node - it solves problem for some time.

Automated calico restart (valid for rke cluster):
```bash
NODENAME=$(hostname -f)
BROKEN=$(for IP in $(KUBECONFIG=/etc/kubernetes/ssl/kubecfg-kube-node.yaml kubectl get pods -A -o jsonpath="{.items[?(@.spec.nodeName==\"$NODENAME\")].status.podIP}"); do ip r get $IP ; done | grep -c bond0)
if [ $BROKEN -gt 0 ]; then
  docker restart $(docker ps | awk '/calico/ {print $1}')
fi
```

## Steps to Reproduce (for bugs)
1. Upgrade calico from 3.13.3 to 3.21.1
2. Wait some time for calico to break.
3. New pods doesn't get functional network.
4. Delete pod and see iptables leftovers.

## Your Environment

* calico 3.21.1 upgraded from 3.13.3
* Kubernetes 1.22.9 upgraded from 1.20.x, installed and managed with rke (currently 1.3.11)
* Ubuntu 20.04

**Comment by mateuszkwiatkowski** [comment]
Update: since updating the cluster to Kubernetes 1.23.6 and calico 3.22.0 problem haven't returned.

**Comment by caseydavenport** [comment]
@mateuszkwiatkowski how large was your cluster? There was a bug fixed in v3.21.3 and onward that had similar symptoms to what you are describing, for clusters with >500 of any API resource (e.g., pods, policies, etc).

> Fix a serious regression introduced in v3.21 where the datastore watcher could get stuck and report stale information in clusters with >500 policies/pods/etc. The bug was triggered by needing to do a resync (for example after an etcd compaction) when there were enough resources to trigger the list pager. [calico #5324](https://github.com/projectcalico/calico/pull/5324) (@fasaxc)

**Comment by mateuszkwiatkowski** [comment]
Hi @caseydavenport,

Thank you for reply. We have almost 1k pods so I think this is it! I'm going to close this issue as it didn't happen again after the upgrade.
